File: src/lib/midi.ts
// Minimal Standard MIDI File parser — enough to drive the lessons highway.
//
// A lesson MIDI may hold several tracks. Their role is decided by the track
// name (FF 03 meta) and channel:
//   - "drums", or any track on the percussion channel (10) -> PLAYABLE:
//     these notes are shown on the highway and scored.
//   - "family:id" (e.g. "bass:lately")               -> BACKING:
//     auto-played from static/<family>/<id>/<note>.oga, never shown or scored.
//   - "count-in"                                     -> COUNT-IN:
//     the stick clicks that lead the student in — audible on the lesson's own
//     kit, never shown or scored. Checked before the percussion-channel rule,
//     which it would otherwise trip.
//   - "guide"                                        -> GUIDE:
//     the hi-hat a hatless lesson borrows to keep time against — audible on the
//     lesson's own kit, never shown or scored. Same exemption from the
//     percussion-channel rule as the count-in, and for the same reason.

/** Beats of lead-in before a lesson's beat 0 — the bar the count-in lives in. */
export const COUNT_IN_BEATS = 4;

/**
 * How long the transport keeps running past the last note in the file.
 *
 * One beat: enough for the closing hit to sit inside its own match window and
 * be scored on its merits, and for the bass's resolving tonic to ring before
 * the result screen arrives. Patterns whose last note is not on a bar line are
 * unaffected — bar rounding already carries them further than this.
 */
const TAIL_BEATS = 1;

// `vel` is MIDI velocity 1-127, carried only where it is used: backing tracks
// play it as loudness. Playable notes are scored on timing alone, so nothing
// on that path reads it.
export type MidiNote = { beat: number; note: number; vel?: number };
export type BackingTrack = { family: string; id: string; notes: MidiNote[] };

export type ParsedMidi = {
  ppq: number;
  bpm: number;
  notes: MidiNote[]; // playable (scored) notes
  backing: BackingTrack[]; // auto-played accompaniment
  /**
   * Count-in clicks, already shifted onto the transport's beat axis: the track
   * is written in its own lead-in bar, so these come back at negative beats
   * (-3, -2, -1). Beat 0 is deliberately silent — it belongs to the student.
   */
  countIn: MidiNote[];
  /**
   * The borrowed hi-hat, on the pattern's own beat axis (0 upward). Present
   * only for lessons whose pattern has no hat of its own. Carries `vel`, which
   * the drum player applies as gain so it sits under the student's playing.
   */
  guide: MidiNote[];
  lengthBeats: number;
};

export function parseMidi(buf: ArrayBuffer): ParsedMidi {
  const d = new DataView(buf);
  let p = 0;

  const tag = () =>
    String.fromCharCode(
      d.getUint8(p++),
      d.getUint8(p++),
      d.getUint8(p++),
      d.getUint8(p++),
    );
  const u32 = () => {
    const v = d.getUint32(p);
    p += 4;
    return v;
  };
  const u16 = () => {
    const v = d.getUint16(p);
    p += 2;
    return v;
  };

  if (tag() !== "MThd") throw new Error("Not a MIDI file");
  u32(); // header length (always 6)
  u16(); // format
  const ntracks = u16();
  const ppq = u16(); // assume metrical timing (positive)

  function varlen() {
    let v = 0;
    for (;;) {
      const b = d.getUint8(p++);
      v = (v << 7) | (b & 0x7f);
      if (!(b & 0x80)) break;
    }
    return v;
  }

  const playable: MidiNote[] = [];
  const backing: BackingTrack[] = [];
  let countIn: MidiNote[] = [];
  let guide: MidiNote[] = [];
  let bpm = 120;
  let maxTick = 0;

  for (let t = 0; t < ntracks; t++) {
    if (tag() !== "MTrk") break;
    const len = u32();
    const end = p + len;
    let tick = 0;
    let running = 0;
    let name = "";
    let usesPercussion = false;
    let trackMax = 0;
    const trackNotes: MidiNote[] = [];

    while (p < end) {
      tick += varlen();
      let statusByte = d.getUint8(p);
      if (statusByte & 0x80) p++;
      else statusByte = running; // running status
      running = statusByte;

      const cmd = statusByte & 0xf0;
      const chan = statusByte & 0x0f;
      if (statusByte === 0xff) {
        const type = d.getUint8(p++);
        const mlen = varlen();
        if (type === 0x03) {
          let s = "";
          for (let k = 0; k < mlen; k++)
            s += String.fromCharCode(d.getUint8(p + k));
          name = s;
        } else if (type === 0x51 && mlen === 3) {
          const us =
            (d.getUint8(p) << 16) |
            (d.getUint8(p + 1) << 8) |
            d.getUint8(p + 2);
          bpm = Math.round(60_000_000 / us);
        }
        p += mlen;
      } else if (statusByte === 0xf0 || statusByte === 0xf7) {
        p += varlen();
      } else if (cmd === 0x90) {
        const note = d.getUint8(p++);
        const vel = d.getUint8(p++);
        if (chan === 9) usesPercussion = true;
        if (vel > 0) trackNotes.push({ beat: tick / ppq, note, vel });
        if (tick > trackMax) trackMax = tick;
      } else if (cmd === 0x80 || cmd === 0xa0 || cmd === 0xb0 || cmd === 0xe0) {
        p += 2;
      } else if (cmd === 0xc0 || cmd === 0xd0) {
        p += 1;
      } else {
        p++; // unknown — skip a byte defensively
      }
    }
    p = end;

    if (!trackNotes.length) continue;

    // The count-in lives in the bar BEFORE the pattern, so it neither counts
    // toward the pattern's length nor belongs on the percussion/playable path.
    if (name.trim() === "count-in") {
      countIn = trackNotes
        .map((n) => ({ beat: n.beat - COUNT_IN_BEATS, note: n.note }))
        .sort((a, b) => a.beat - b.beat);
      continue;
    }

    // The guide runs the length of the pattern and no further, so it cannot
    // extend the lesson — but it is still percussion on channel 10, so it has
    // to be claimed here or the rule below would put it on the highway.
    if (name.trim() === "guide") {
      guide = trackNotes.sort((a, b) => a.beat - b.beat);
      continue;
    }

    if (trackMax > maxTick) maxTick = trackMax;
    const colon = name.indexOf(":");
    if (colon > 0 && !usesPercussion) {
      // backing track: "family:id"
      backing.push({
        family: name.slice(0, colon).trim(),
        id: name.slice(colon + 1).trim(),
        notes: trackNotes.sort((a, b) => a.beat - b.beat),
      });
    } else {
      playable.push(...trackNotes);
    }
  }

  playable.sort((a, b) => a.beat - b.beat);
  const lastBeat = maxTick / ppq;
  // Round the length up to a whole bar (4 beats) so the loop stays musical.
  const barRounded = Math.max(4, Math.ceil(lastBeat / 4) * 4);
  // A lesson closes ON the bar line that follows it — the pattern's down-beat
  // struck once more, with the bass resolving onto the same beat. Bar rounding
  // alone would end the transport on that very beat: the closing note would
  // fall outside its own match window and be scored a miss no matter how well
  // it was played, and the resolution would be cut off as the report appeared.
  // So the run always outlasts its last note.
  const lengthBeats = Math.max(barRounded, lastBeat + TAIL_BEATS);
  return { ppq, bpm, notes: playable, backing, countIn, guide, lengthBeats };
}

File: scripts/lessons/midi.py
"""MIDI mechanics: bytes in, file out. No curriculum knowledge lives here.

A lesson MIDI is format 1, one track per role, and the track *name* is what
tells the app what to do with it:

  "drums"     playable: shown on the highway and scored.
  "family:id" backing:  auto-played from static/<family>/<id>/<note>.oga,
              never shown, never scored (e.g. "bass:lately").
  "count-in"  the stick count that leads the student in; audible, never scored.
"""

import struct

PPQ = 480
BEATS_PER_BAR = 4

# GM percussion notes the curriculum uses. The count-in clicks on the side stick:
# a dry rim sound that cuts through a kit without being mistaken for a pad the
# student is about to play. (GM 31 "Sticks" would read better still, but
# render-drums.py only renders 35-70.)
KICK, SNARE, CLOSED_HH, OPEN_HH, SIDE_STICK = 36, 38, 42, 46, 37


def varint(n):
    out = [n & 0x7F]
    n >>= 7
    while n:
        out.insert(0, (n & 0x7F) | 0x80)
        n >>= 7
    return bytes(out)


def build_track(name, events, length_ticks, meta=None):
    """events: list of (tick, order, raw_bytes); order sorts ties (offs<ons)."""
    ev = list(meta or [])
    if name:
        nm = name.encode("latin1")
        ev.append((0, -10, bytes([0xFF, 0x03]) + varint(len(nm)) + nm))
    allev = sorted(ev + events, key=lambda e: (e[0], e[1]))
    body = b""
    prev = 0
    for tick, _order, raw in allev:
        body += varint(tick - prev) + raw
        prev = tick
    body += varint(max(0, length_ticks - prev)) + bytes([0xFF, 0x2F, 0x00])
    return b"MTrk" + struct.pack(">I", len(body)) + body


def write_midi(path, tracks):
    hdr = b"MThd" + struct.pack(">IHHH", 6, 1, len(tracks), PPQ)
    with open(path, "wb") as f:
        f.write(hdr + b"".join(tracks))


def hit(events, tick, note):
    """A percussion strike on channel 10: note-on then a short note-off."""
    events.append((tick, 1, bytes([0x99, note, 100])))
    events.append((tick + 60, 0, bytes([0x89, note, 0])))


def bass_note(events, tick, note, dur=200):
    """A bass note on channel 1."""
    events.append((tick, 1, bytes([0x90, note, 95])))
    events.append((tick + dur, 0, bytes([0x80, note, 0])))


def count_in_sticks():
    """The stick count that leads the student in — three clicks, then silence.

    Written in a lead-in bar of its own: beats 2, 3 and 4 of that bar click, and
    the bar line that follows IS the pattern's first beat. The app plays this
    track one bar early (it already holds an empty bar of lead-in before the
    pattern), so the count runs "click, click, click" and the student's own first
    hit lands on the empty down-beat that follows.
    """
    events = []
    for beat in (1, 2, 3):  # the 4th beat is left silent — the down-beat is theirs
        hit(events, beat * PPQ, SIDE_STICK)
    return events, BEATS_PER_BAR * PPQ
